### ISSUE TITLE
[FEATURE] Smart retry to avoid "rate_limit_exceeded" using OpenAiChatModel

### ISSUE THREAD

**Issue by alwa** [langchain4j/langchain4j#814]
I get this kind of error for some queries using OpenAiChatModel:

```
dev.ai4j.openai4j.OpenAiHttpException: {
    "error": {
        "message": "Rate limit reached for gpt-4-0613 in organization <MASKED> on tokens per min (TPM): Limit 10000, Used 5775, Requested 4301. Please try again in 456ms. Visit https://platform.openai.com/account/rate-limits to learn more.",
        "type": "tokens",
        "param": null,
        "code": "rate_limit_exceeded"
    }
}
```

There is a "maxRetries" option but retries are immediately executed. Is there a way to add delay between retries?

**Comment by BlueJoneleth** [comment]
https://platform.openai.com/docs/guides/rate-limits/usage-tiers?context=tier-four

If your company has the budget for it, the easiest thing to do is to charge your account. Your tier gets upgraded the more you spent and you can get to tier 4 permanently by spending 250$, which gives you a 300K/min limit on gpt 4

**Comment by alwa** [comment]
> https://platform.openai.com/docs/guides/rate-limits/usage-tiers?context=tier-four
> 
> If your company has the budget for it, the easiest thing to do is to charge your account. Your tier gets upgraded the more you spent and you can get to tier 4 permanently by spending 250$, which gives you a 300K/min limit on gpt 4

Hi, this is for a private project. I'm reluctant to spend that amount but thanks for the advice

**Comment by BlueJoneleth** [comment]
You can always switch to 3.5 it has a higher token limit than 4.

**Comment by alwa** [comment]
> You can always switch to 3.5 it has a higher token limit than 4.

Yes, but it is also awful, relatively speaking 🙂

**Comment by edeandrea** [comment]
You could add some resilience to your project ([Resilience4J](https://github.com/resilience4j/resilience4j) / [MicroProfile fault tolerance](https://download.eclipse.org/microprofile/microprofile-fault-tolerance-4.0.2/microprofile-fault-tolerance-spec-4.0.2.html) / [Bucket4J](https://bucket4j.com/) or something like that) by adding a circuit breaker, bulkhead, rate limit, and fallback.

**Comment by alwa** [comment]
Thanks @edeandrea,
I'll consider it

**Comment by langchain4j** [comment]
@alwa yes, I think we can add this logic to retry after period specified in the response.
Retry logic is pretty dumb now.

**Comment by alwa** [comment]
> @alwa yes, I think we can add this logic to retry after period specified in the response. Retry logic is pretty dumb now.

Should I report this as a separate feature request?

**Comment by langchain4j** [comment]
@alwa no need, let's keep this one

**Comment by seinecle** [comment]
Upvoting
